### PR TITLE
test: Add a LB simulation test to validate selection probabilities

### PR DESCRIPTION
**Pull request by tonya11en**
Adds a LB simulation test to ensure that any subtle behavioral changes to the Least Request LB will cause a test failure. The test verifies the selection probabilities are as expected in various contexts. 

The intent of this test is **not** to ossify the behavior of the LB, but to flag any potential unintended behavioral changes.

This test was run 10,000 times (using remote executors) to check for flakiness and there was not any observed failure:
```
 txallen@stormbeard  ~/src/envoy/test/common/upstream   lb-sim  bazel test //test/common/upstream:load_balancer_simulation_test --runs_per_test 10000
INFO: Invocation ID: 994289cb-9172-4db7-9307-30f202507d87
INFO: Analyzed target //test/common/upstream:load_balancer_simulation_test (0 packages loaded, 527 targets configured).
INFO: Found 1 test target...
Target //test/common/upstream:load_balancer_simulation_test up-to-date:
  bazel-bin/test/common/upstream/load_balancer_simulation_test
INFO: Elapsed time: 87.417s, Critical Path: 21.53s
INFO: 10001 processes: 1 internal, 10000 remote.
INFO: Build completed successfully, 10001 total actions

Executed 1 out of 1 test: 1 test passes.
```

### CODE REVIEW

**Comment by tonya11en** [comment]
@adisuissa @wbpcode this PR should now be ready for a first pass.

File: changelogs/current.yaml [hunk 1]
```diff
@@ -159,6 +159,11 @@ new_features:
   change: |
     Added :ref:`the Basic Auth filter <envoy_v3_api_msg_extensions.filters.http.basic_auth.v3.BasicAuth>`, which can be used to
     authenticate user credentials in the HTTP Authentication heaer defined in `RFC7617 <https://tools.ietf.org/html/rfc7617>`_.
+- area: upstream
+  change: |
+    Added support to use http async client to fetch the credentials from EC2 instance metadata and ECS task metadata providers
+    instead of libcurl which is deprecated. By default this behavior is disabled. To enable set
+    ``envoy.reloadable_features.use_http_client_to_fetch_aws_credentials`` to true.
```
Comment (adisuissa) [inline changelogs/current.yaml:166]: Doesn't seem related to the current PR, should probably be reverted.

File: docs/root/intro/arch_overview/upstream/load_balancing/load_balancers.rst [hunk 2]
```diff
@@ -36,9 +36,9 @@ same or different weights.
   host which has the fewest active requests (`Mitzenmacher et al.
   <https://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf>`_ has shown that this
   approach is nearly as good as an O(N) full scan). This is also known as P2C (power of two
-  choices). The P2C load balancer has the property that a host with the highest number of active
-  requests in the cluster will never receive new requests. It will be allowed to drain until it is
-  less than or equal to all of the other hosts.
+  choices). The P2C load balancer has the property that host weights will decrease as the number of
+  active requests on those hosts increases. P2C selection is particularly useful for load
+  balancer implementations due to its resistance to herding behavior.
```
Comment (adisuissa) [inline docs/root/intro/arch_overview/upstream/load_balancing/load_balancers.rst:41]: nit: consider defining/giving example of "herding behavior".

File: test/common/upstream/load_balancer_simulation_test.cc [hunk 3]
```diff
@@ -81,26 +113,88 @@ TEST(DISABLED_LeastRequestLoadBalancerWeightTest, Weight) {
...
+  params.tolerance_pct = 1.0;
+  leastRequestLBWeightTest(params);
+
+  // Same as above, but with 3 hosts.
+  params.num_hosts = 3;
+  params.num_subset_hosts = 1;
+  params.active_request_count = 1;
+  params.expected_selection_probs = {33.3, 33.3, 33.3};
+  params.weight = 2;
+  params.tolerance_pct = 1.0;
```
Comment (adisuissa) [inline test/common/upstream/load_balancer_simulation_test.cc:194]: Seems this value is always 1.0. Consider removing it from being a parameter, and set it to a const in the function.

File: test/common/upstream/load_balancer_simulation_test.cc [hunk 4]
```diff
@@ -81,26 +113,88 @@ TEST(DISABLED_LeastRequestLoadBalancerWeightTest, Weight) {
       priority_set, nullptr, lb_stats, runtime, random, common_config, least_request_lb_config,
       *time_source};
 
-  absl::node_hash_map<HostConstSharedPtr, uint64_t> host_hits;
-  const uint64_t total_requests = 100;
-  for (uint64_t i = 0; i < total_requests; i++) {
+  for (uint64_t i = 0; i < n; i++) {
```
Comment (adisuissa) [inline test/common/upstream/load_balancer_simulation_test.cc:116]: I find it strange, as `n` is no longer the number of `total_requests`.
Can you add some comments clarifying what is the intention here?

**Review by adisuissa — COMMENTED** [review]
Thanks!
Minor comments, but otherwise LGTM!

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
